src/languages/php.js

```diff
@@ -263,13 +264,76 @@ export default function(hljs) {
     "stdClass"
   ];
 
+  /** Dual-case keywords
+   *
+   * ["then","FILE"] =>
+   *     ["then", "THEN", "FILE", "file"]
+   *
+   * @param {string[]} items */
+  const dualCase = (items) => {
+    /** @type string[] */
+    const result = [];
+    items.forEach(item => {
+      result.push(item);
+      if (item.toLowerCase() === item) {
+        result.push(item.toUpperCase());
+      } else {
+        result.push(item.toLowerCase());
+      }
+    });
+    return result;
+  };
+
   const KEYWORDS = {
     keyword: KWS,
-    literal: LITERALS,
-    built_in: BUILT_INS
+    literal: dualCase(LITERALS),
+    built_in: BUILT_INS,
+  };
+
+  /**
+   * @param {string[]} items */
+  const normalizeKeywords = (items) => {
+    return items.map(item => {
+      return item.replace(/\|\d+$/, "");
+    });
+  };
+
+  const CONSTRUCTOR_CALL = {
+    variants: [
+      {
+        match: [
+          /new/,
+          /\s+/,
+          // to prevent built ins from being confused as the class constructor call
+          regex.concat("(?!", normalizeKeywords(BUILT_INS).join("\\b|"), "\\b)"),
+          /\\?\w+/,
+          /\s*\(/,
+        ],
+        scope: {
+          1: "keyword",
+          4: "title.class",
+        },
+      }
+    ]
   };
+
+  const FUNCTION_INVOKE = {
+    relevance: 0,
+    match: [
+      /(?:->|::|\s|\(|\\)/,
```

Comment:
Please provide commented examples of all the variants you're using here. Is just an attempt to eliminate false positives - if so what would an example or two be?   Why do we need these at all vs just relying on `[ident](`...?

Reply:
> Is just an attempt to eliminate false positives - if so what would an example or two be? Why do we need these at all vs just relying on `[ident](`...?
> 
> If this is critical please provide commented code examples of all the variants you're using here.

All the correct cases are in the `Function invoke` tests. I also added a comment and test for `First-class Callable Syntax`,
which was working correctly before this PR (because there was no function invoke). Now it would be a false positive, so I added all the rules.

Reply:
I'm not following...   Isn't this just an issue with the non-function guard?  If I do this, all tests pass:

```
diff --git a/src/languages/php.js b/src/languages/php.js
index 9f45841a..297bd55b 100644
--- a/src/languages/php.js
+++ b/src/languages/php.js
@@ -321,7 +321,8 @@ export default function(hljs) {
     relevance: 0,
     match: [
       // to prevent First-class Callable Syntax from being confused as the function title
-      /(?:->|::|\s|\(|\\)/,
+      // /(?:->|::|\s|\(|\\)/,
+      /\b/,
       // to prevent keywords from being confused as the function title
       regex.concat("(?!fn\\b|function\\b|", normalizeKeywords(KWS).join("\\b|"), "|", normalizeKeywords(BUILT_INS).join("\\b|"), "\\b)"),
       /\w+/,
@@ -375,7 +376,8 @@ export default function(hljs) {
       FUNCTION_INVOKE,
       {
         // swallow composed identifiers to avoid parsing them as keywords
-        begin: /(::|->)+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*/
+        begin: /(::|->)+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*(?!\s*\()(?![a-zA-Z0-9_\x7f-\xff])/,
+        // scope:"wrong"
       },
       CONSTRUCTOR_CALL,
       {
```

IE, fixing the guard and using only a word boundary to match the start of function invokes... thoughts?  This using multi-match as "look-behind" is dangerous and I really don't want to encourage it at all if it can be avoided.

Reply:
@joshgoebel Strange because I applied your fix and they fall.
```
1 failing

  1) php
       should not cause polynomial backtracking:
     Error: php/[10]/begin: Polynomial backtracking. By repeating any character that matches /\xa0/i, an attack string can be created.

    [a-zA-Z0-9_\x7f-\xff]*(?!\s*
    ^~~~~~~~~~~~~~~~~~~~~~   ^~~

Full pattern:
/(::|->)+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*(?!\s*\()(?![a-zA-Z0-9_\x7f-\xff])/m

```

After changing `\s*` to `\s` I have a below error and a few offers.

```
-DateTimeImmutable::createFromMutable(<span class="hljs-keyword">new</span> <span class="hljs-title class_">\DateTime</span>(<span class="hljs-string">&#x27;now&#x27;</span>));
      +DateTimeImmutable::<span class="hljs-function title_ invoke__">createFromMutable</span>(<span class="hljs-keyword">new</span> <span class="hljs-title class_">\DateTime</span>(<span class="hljs-string">&#x27;now&#x27;</span>));
```
This regex do not detect static function invoke:
`DateTimeImmutable::createFromMutable();`
or method invoke
`$date->format();`
and we do not want to higlight First-class Callable Syntax as a function
`$test();`

P.S. I'm run tests inside Docker.

Reply:
> should not cause polynomial backtracking:

That can be ignored [for now] and fixed.  **I didn't have any markup test errors.**  Your'e saying you're seeing the same result right - all tests pass , other than the security issue?

So if we can fix that, then can this be simplified as I'm suggesting?

Reply:
Can names TRULY contain non-breaking spaces (0xA0)? That sounds like a bug to me...

Reply:
> > should not cause polynomial backtracking:
> 
> That can be ignored [for now] and fixed. **I didn't have any markup test errors.** Your'e saying you're seeing the same result right - all tests pass with my patch , other than the security issue?
> 
> So if we can fix that, then can this be simplified as I'm suggesting?

It will be annoying



> Can names TRULY contain non-breaking spaces (0xA0)? That sounds like a bug to me...

[Yes](https://mnapoli.fr/using-non-breakable-spaces-in-test-method-names/), but I never saw anyone using it. Looks bad for me.

Reply:
@joshgoebel fixed it with ` *` because only space is valid before the name of the function and opening bracket.